### PR TITLE
Sql task Dzianis Shupenka

### PR DESCRIPTION
**Pull request by dzianisshupenka**
https://travis-ci.org/github/dzianisshupenka/db-assignments/builds/714782982

### CODE REVIEW

File: task/sql-tasks.js [hunk 1]
```diff
@@ -67,7 +84,18 @@ async function task_1_3(db) {
...
 async function task_1_4(db) {
-    throw new Error("Not implemented");
+
+    let result = await db.query(`
+    SELECT 
+        CustomerID AS 'Customer Id',
+        COUNT(OrderID) AS 'Total number of Orders',
+        ROUND((COUNT(OrderID)/(SELECT COUNT(*) AS 'All' from Orders)) * 100, 5) AS '% of all orders'
+        FROM Orders
+    GROUP BY Orders.CustomerID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: if you're only 1 table is used in the query no need to write it in from of the column

File: task/sql-tasks.js [hunk 2]
```diff
@@ -91,7 +128,16 @@ async function task_1_5(db) {
...
  */
 async function task_1_6(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT 
+        Products.ProductName,
+        Categories.CategoryName,
+        Suppliers.CompanyName AS 'SupplierCompanyName'
+    FROM Products JOIN Categories ON Products.CategoryID = Categories.CategoryID
+    JOIN Suppliers ON Products.SupplierID = Suppliers.SupplierID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use **inner join** keyword explicitly

File: task/sql-tasks.js [hunk 3]
```diff
@@ -105,7 +151,17 @@ async function task_1_6(db) {
  *
  */
 async function task_1_7(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT 
+        subEmployeers.EmployeeID AS 'EmployeeId',
+        CONCAT(subEmployeers.FirstName, ' ', subEmployeers.LastName) AS 'FullName',
+        IF(subEmployeers.ReportsTo IS NULL, '-' , CONCAT(mainEmployeers.FirstName, ' ', mainEmployeers.LastName))  AS 'ReportsTo'
+    FROM Employees AS subEmployeers
```
Comment (AlexMokin) [inline task/sql-tasks.js]: for short queries short aliases a preferred. E.g. empl and subEmpl

File: task/sql-tasks.js [hunk 4]
```diff
@@ -129,7 +193,14 @@ async function task_1_8(db) {
  *
  */
 async function task_1_9(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            CustomerID,
+            ContactName
+        FROM Customers
+        WHERE ContactName RLIKE "^F..n"
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use like if possible

File: task/sql-tasks.js [hunk 5]
```diff
@@ -177,7 +272,14 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
```
Comment (AlexMokin) [inline task/sql-tasks.js:275]: it's better to write this query like this
notice that from-clause is optional.



```
select
   (select count()...) as 'T'
   (select count()...) as 'TT'
```

File: task/sql-tasks.js [hunk 6]
```diff
@@ -199,7 +309,24 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            SUM(MONTH(OrderDate) = 1) AS 'January',
+            SUM(MONTH(OrderDate) = 2) AS 'February',
+            SUM(MONTH(OrderDate) = 3) AS 'March',
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count records, it shows intent better

File: task/sql-tasks.js [hunk 7]
```diff
@@ -259,7 +423,19 @@ async function task_1_19(db) {
...
 async function task_1_20(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            Employees.EmployeeID,
+            CONCAT(FirstName, ' ', LastName) AS 'Employee Full Name',
+            SUM(OrderDetails.UnitPrice * OrderDetails.Quantity) AS 'Amount, $'
+        FROM Employees
+        JOIN Orders ON Orders.EmployeeID = Employees.EmployeeID
+        JOIN OrderDetails ON Orders.OrderID = OrderDetails.OrderID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use **inner join** keyword explicitly

File: task/sql-tasks.js [hunk 8]
```diff
@@ -280,7 +465,26 @@ async function task_1_21(db) {
...
+        FROM Customers
+        JOIN Orders ON Orders.CustomerID = Customers.CustomerID
+        JOIN OrderDetails ON Orders.OrderID = OrderDetails.OrderID
+        JOIN Products ON Products.ProductID = OrderDetails.ProductID
+        WHERE OrderDetails.UnitPrice = (
+            SELECT 
+                MAX(OrderDetails.UnitPrice) FROM Customers AS customers2
+                JOIN Orders ON Customers.CustomerID = Orders.CustomerID
+                JOIN OrderDetails ON Orders.OrderID = OrderDetails.OrderID
+                WHERE customers2.CompanyName = Customers.CompanyName
```
Comment (AlexMokin) [inline task/sql-tasks.js]: company name doesn't uniquely identify customer. also this column doesn't have an index so it's also slow.

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js (hunk 1 above)
Comment (dzianisshupenka) [reply task/sql-tasks.js]: Sorry, I don't understand what is the problem. What do I need to do?

**Comment by dzianisshupenka** [comment]
https://travis-ci.org/github/dzianisshupenka/db-assignments/builds/724486964

File: task/sql-tasks.js [hunk 9]
```diff
@@ -67,7 +84,18 @@ async function task_1_3(db) {
...
 async function task_1_4(db) {
-    throw new Error("Not implemented");
+
+    let result = await db.query(`
+    SELECT 
+        CustomerID AS 'Customer Id',
+        COUNT(OrderID) AS 'Total number of Orders',
+        ROUND((COUNT(OrderID)/(SELECT COUNT(*) AS 'All' FROM Orders)) * 100, 5) AS '% of all orders'
+    FROM Orders
+    GROUP BY Orders.CustomerID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: you should specify only column names if you select from a single table 

image

File: task/sql-tasks.js [hunk 10]
```diff
@@ -280,7 +463,26 @@ async function task_1_21(db) {
...
+        SELECT DISTINCT
+            Customers.CompanyName,
+            Products.ProductName,
+            OrderDetails.UnitPrice AS 'PricePerItem'
+        FROM Customers
+        INNER JOIN Orders ON Orders.CustomerID = Customers.CustomerID
+        INNER JOIN OrderDetails ON Orders.OrderID = OrderDetails.OrderID
+        INNER JOIN Products ON Products.ProductID = OrderDetails.ProductID
+        WHERE OrderDetails.UnitPrice = (
+            SELECT
```
Comment (AlexMokin) [inline task/sql-tasks.js]: still kinda slow, try to use group by in this subquery and join this subquery on customer id

**Comment by AlexMokin** [comment]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by dzianisshupenka** [comment]
I fixed all previous comments, what I need to do now?

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by dzianisshupenka** [comment]
https://travis-ci.org/github/dzianisshupenka/db-assignments/builds/725497782

**Comment by AlexMokin** [comment]
accepted

**Pull request closed without merging**